pbft/pbft.cpp

```diff
@@ -1041,13 +1056,443 @@ pbft::get_high_water_mark()
+
+    return retval;
+}
+
+bool
+pbft::is_valid_viewchange_message(const pbft_msg& viewchange_message, const bzn_envelope& original_msg) const
+{
+    if (!this->is_peer(original_msg.sender()))// TODO: Rich - If redundant, keep this check, remove the other
+    {
+        LOG(debug) << "is_valid_viewchange_message - message not from a peer";
+        return false;
+    }
+
+    if (!(viewchange_message.view() > this->get_view()))
+    {
+        LOG(error) << "is_valid_viewchange_message - new view " <<  viewchange_message.view() << " is not greater than current view " << this->get_view();
+        return false;
+    }
+
+    auto valid_checkpoint_hashes = this->validate_and_extract_checkpoint_hashes(viewchange_message);
+
+    if (valid_checkpoint_hashes.empty())
+    {
+        LOG(error) << "is_valid_viewchange_message - the checkpoint is invalid";
+        return false;
+    }
+    const auto valid_checkpoint = *(valid_checkpoint_hashes.begin());
+
+
+    // TODO: (see KEP-882) Refactor this block into a new method - Isabel: should extract a validate_prepared_proof method
+    // all the the prepared proofs are valid  (contains a pre prepare and 2f+1 mathcin prepares)
+    for (int i{0}; i < viewchange_message.prepared_proofs_size(); ++i)
+    {
+        const bzn_envelope& pre_prepare_envelope{viewchange_message.prepared_proofs(i).pre_prepare()};
+        if (!this->is_peer(pre_prepare_envelope.sender()) || !this->crypto->verify(pre_prepare_envelope))
+        {
+            LOG(error) << "is_valid_viewchange_message - a pre prepare message has a bad envelope, or the sender is not in the peers list";
+            return false;
+        }
+
+        pbft_msg preprepare_message;
+        if (!preprepare_message.ParseFromString(pre_prepare_envelope.pbft()) || (preprepare_message.sequence() <= valid_checkpoint.first.first))
+        {
+            LOG(error) << "is_valid_viewchange_message - a pre prepare message has an invalid sequence number, or is malformed";
+            return false;
+        }
+
+        std::set<uuid_t> senders;
+        for (int j{0}; j < viewchange_message.prepared_proofs(i).prepare_size(); ++j)
+        {
+            bzn_envelope prepare_envelope{viewchange_message.prepared_proofs(i).prepare(j)};
+            if (!this->is_peer(prepare_envelope.sender()) || !this->crypto->verify(prepare_envelope))
+            {
+                LOG(error) << "is_valid_viewchange_message - a prepare message has a bad envelope, the sender may not be in the peer list, or the envelope failed cryptographic verification";
+                return false;
+            }
+
+            // does the sequence number, view number and hash match those of the pre prepare
+            pbft_msg prepare_msg;
+            if (!prepare_msg.ParseFromString(prepare_envelope.pbft()))
+            {
+                LOG(error) << "is_valid_viewchange_message - a prepare message is invalid";
+                return false;
+            }
+
```

Comment:
check that the prepare has correct type.

Reply:
I hope that the correct type is PBFT_MSG_VIEWCHANGE and *not* PBFT_MSG_PREPARE

I am now checking for PBFT_MSG_VIEWCHANGE, is that correct?

if (!prepare_msg.ParseFromString(prepare_envelope.pbft()) || prepare_msg.type() == PBFT_MSG_VIEWCHANGE)